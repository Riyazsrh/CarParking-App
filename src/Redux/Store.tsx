import { createStore } from "redux"
import handleReducer from "./Reducer"

const Store = createStore(handleReducer);


export default Store;
