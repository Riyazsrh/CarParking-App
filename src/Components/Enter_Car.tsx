import { Button, Card, CardContent, Grid, Paper, TextField, Typography } from '@mui/material'
import { Box } from '@mui/system';
import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { useNavigate } from 'react-router-dom';
import { handleId, handleTextId, HANDLE_ID, HANDLE_Text_ID } from '../Redux/Actoin';
import { motion } from "framer-motion"
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';

const Enter_Car = () => {
  const carSlots = useSelector((state: any) => state.carSlots);
  const ids = useSelector((state: any) => state.id);
  const [text, setText] = useState<any>();
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const handleChange = (e: any) => {
    setText(e.target.value)
  }

  const handleCarSlots = () => {
    if (!text) {
      toast.dark("please enter anything")
    } else {

      let temp = [...carSlots]

      let emptyslots = temp.filter((item: any) => item.EnterCarNo === '')
      if (emptyslots.length === 0) {
        toast('Slot is not empty')
      } else {

        //random_slot
        const id = emptyslots[Math.floor(Math.random() * emptyslots.length)].id
        console.log('iddddd', id);

        const indexno = temp.findIndex((i) => i.id === id)
        console.log('szdfg', indexno);

        const current = new Date();
        const time = `${current.getHours()}:${current.getMinutes()}`;
        temp[indexno].available = false;
        temp[indexno].EnterCarNo = text;
        temp[indexno].carParkingTime = time;  //new Date().setHours(new Date().getHours() - 3);

        dispatch(handleId(HANDLE_ID, temp))
        console.log('length', temp);

        setText('')
      }
    }
  }

  const handleExit = (id: any) => {
    dispatch(handleTextId(HANDLE_Text_ID, id))
    console.log("exitItem", id);

    navigate('/exitCar')
  }

  return (
    <motion.div
      initial={{ width: 0 }}
      animate={{ width: "100%" }}
      exit={{ x: window.innerWidth }}
    >
      <Grid >
        <Grid style={{ display: 'flex', justifyContent: "center", marginTop: 50 }}>
          <Button color='warning' onClick={() => { navigate('/') }}><ArrowBackIcon /></Button>
          <TextField data-testid="Car_Input_Check" onChange={handleChange} value={text} name='text' placeholder='Enter Car no' style={{ width: 350 }} />
          <Button data-testid="CarNo_Button_Check" onClick={handleCarSlots}>Add Car</Button>
        </Grid>

        <Box style={{ gap: "15px", height: 80, }}>
          {
            carSlots?.map((item: any, i: any) => {
              return (
                <Grid container justifyContent='center'>
                  <Grid item xs={6} >
                    <Card onClick={() => !item.EnterCarNo ? toast.warning('This CarSlot is empty') : handleExit(item)} key={i} style={{ margin: '20px', cursor: "pointer", background: !item.EnterCarNo ? '' : '#FFEDDB' }}>
                      <CardContent>
                        <Typography color="textSecondary" gutterBottom>
                          id: <Box display='inline'  > {item.id} </Box>
                        </Typography>
                        <Typography variant="h6" component="h2">
                          CarNo:   <Box display='inline' fontWeight='900'  >{item.EnterCarNo}</Box>
                        </Typography>
                        {!item.EnterCarNo ?
                          null : <Typography marginTop="5px" >
                            Time:  <Box display='inline' fontWeight='900'  >{item.carParkingTime}</Box>
                          </Typography>}
                      </CardContent>
                    </Card>
                  </Grid>
                </Grid>
              )
            })
          }
        </Box>
      </Grid>
      <ToastContainer position="top-center"
        autoClose={2000}
        hideProgressBar={false}
        closeOnClick
        rtl={false} />
    </motion.div>
  )
}

export default Enter_Car
