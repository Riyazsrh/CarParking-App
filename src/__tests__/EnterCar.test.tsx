import { act, fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import Enter_Car from "../Components/Enter_Car";
import Store from "../Redux/Store";




const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
    useSelector: jest.fn(),
    useDispatch: () => mockDispatch
}));


const mockUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => mockUsedNavigate,
}));


const data = [
    {
        id: 1,
        EnterCarNo: "1232",
        carParkingTime: new Date(),
        available: true,
    },
];

const value: any = {
    carSlots: data,
    ids: "",

};


// test("should fireEvent checks here", () => {

//     const { getByPlaceholderText } = render(<Provider store={value}><Enter_Car /></Provider>);
//     const inputCheck: any = getByPlaceholderText("Enter Car no");

//     act(() => {
//         fireEvent.change(inputCheck, {
//             target: { value: "" }
//         });
//     })

//     expect(inputCheck.value).toBe('');
// });



test("Enter car TextField Ckeck ", () => {
    const doc = render(<Enter_Car />)
    const CarinputCheck = doc.getByTestId('Car_Input_Check')
    expect(CarinputCheck).toBeInTheDocument();
})

test("Enter car  FireEvent check", () => {

    const doc = render(<Enter_Car />);
    const textEvent: any = doc.getByPlaceholderText("Enter Car no");
    const button = doc.getByTestId("CarNo_Button_Check")
    fireEvent.change(textEvent, {
        target: { value: "" }
    });
    fireEvent.click(button)
    expect(textEvent.value).toBe('')
})


test("Add Slots button Check", () => {
    render(<Enter_Car />)
    const button = screen.getByTestId("CarNo_Button_Check")

    fireEvent.click(button);
    expect(button).toBeInTheDocument();

})

test("render submit button on the screen", async () => {
    render(<Enter_Car />);
    const buttonList = await screen.findAllByRole("button");

    expect(buttonList).toHaveLength(2);
});


test("checking snapshot", () => {
    const test1 = render(<Enter_Car />);
    expect(test1).toMatchSnapshot();
});

