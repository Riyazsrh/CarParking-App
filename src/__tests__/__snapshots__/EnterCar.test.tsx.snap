// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checking snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="width: 0px;"
      >
        <div
          class="MuiGrid-root css-vj1n65-MuiGrid-root"
        >
          <div
            class="MuiGrid-root css-vj1n65-MuiGrid-root"
            style="display: flex; justify-content: center; margin-top: 50px;"
          >
            <button
              class="MuiButton-root MuiButton-text MuiButton-textWarning MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-wjixpo-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowBackIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              data-testid="Car_Input_Check"
              style="width: 350px;"
            >
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r4:"
                  name="text"
                  placeholder="Enter Car no"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-hdw1oc"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <button
              class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              data-testid="CarNo_Button_Check"
              tabindex="0"
              type="button"
            >
              Add Car
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiBox-root css-0"
            style="gap: 15px; height: 80px;"
          />
        </div>
        <div
          class="Toastify"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="width: 0px;"
    >
      <div
        class="MuiGrid-root css-vj1n65-MuiGrid-root"
      >
        <div
          class="MuiGrid-root css-vj1n65-MuiGrid-root"
          style="display: flex; justify-content: center; margin-top: 50px;"
        >
          <button
            class="MuiButton-root MuiButton-text MuiButton-textWarning MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-wjixpo-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowBackIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            data-testid="Car_Input_Check"
            style="width: 350px;"
          >
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r4:"
                name="text"
                placeholder="Enter Car no"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-hdw1oc"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            data-testid="CarNo_Button_Check"
            tabindex="0"
            type="button"
          >
            Add Car
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root css-0"
          style="gap: 15px; height: 80px;"
        />
      </div>
      <div
        class="Toastify"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
